### PR TITLE
RFC #1042: Planning the implementation of static evaluation

### PR DESCRIPTION
**Pull request by cam72cam**
This PR is where we will discus how to implement solutions to the problems presented in #1042. This is a living document and will be updated as work progresses.

Goals:
* Share understanding gleaned during the initial prototyping of 1042
* Turn the conclusions drawn from 1042 in to an actionable plan with linked issues
* Get feedback from the community on the plan
* Get signoff from the core-team before implementation starts

### CODE REVIEW

File: docs/static-evaluation.md [hunk 1]
```diff
@@ -0,0 +1,567 @@
+# Implementing Init-time static evaluation of constant variables and locals
+
+As initially described in https://github.com/opentofu/opentofu/issues/1042, many users of OpenTofu are expect to be able to use variables and locals in a variety of locations that are currently not supported. To understand why this is, we need to peek under the hood and understand how and why OpenTofu evaluates expressions in configuration.
+
+## Expressions
+
+The evaluation of expressions (`1 + var.bar` for example) depends on required values and functions used in the expression. In that example, you would need to know the value of `var.bar`. That dependency is known via a concept called "HCL Traversals", which represent an attribute access path and can be turned into strongly typed "OpenTofu References". In practice, you would say "the expression depends on an OpenTofu Variable named bar".
+
+Once you know what the requirements are for an expression (`hcl.Expression`), you can build up an evaluation context (`hcl.EvalContext`) to provide those requirements or return an error.  In the above example, the evaluation context would include `{"var": {"bar": <somevalue>}`.
```
Comment (Yantrio) [inline docs/static-evaluation.md:9]: might be nice to link to the code here (godoc is fine)

File: docs/static-evaluation.md [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+
+Expression evaluation is currently split up into two stages: config loading and reference evaluation
```
Comment (Yantrio) [inline docs/static-evaluation.md:11]:
```suggestion
Expression evaluation is currently split up into two stages: config loading and graph reference evaluation
```

File: docs/static-evaluation.md [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+
+## Config Loading
+
+During configuration loading, the hcl or json config is pulled apart into Blocks and Attributes. A Block contains Attributes and nested Blocks. Attributes are simply named expressions (`foo = 1 + var.bar` for example).
+
+These Blocks and Attributes are abstract representations of the configuration which have not yet been evaluated into actionable values. Depending on the handing of the given block/attribute, either the abstract representation is kept or it is evaluated into a real value for use.
+
+As a concrete example, the `module -> source` field must be known during configuration loading as it is required to continue the next iteration of the loading process.  However attributes like `module -> for_each` may depend on attribute values from resources or other pieces of information not known during config loading and therefore are stored as an expression for later evaluation.
+
+No evaluation context is built or provided during the entire config loading process.  **Therefore, no functions, locals, or variables may be used when during config loading due to the lack of context.  This limitation is what we wish to resolve**.
```
Comment (Yantrio) [inline docs/static-evaluation.md:21]:
```suggestion
No evaluation context is built or provided during the entire config loading process.  **Therefore, no functions, locals, or variables may be used during config loading due to the lack of context.  This limitation is what we wish to resolve**.
```

File: docs/static-evaluation.md [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+
+## Graph Reference Evaluation
+
+After the config is fully loaded, it is transformed and processed into nodes in a graph. These nodes use the "OpenTofu References" present in their blocks/attributes (the ones not evaluated in config loading) to build both the dependency edges in the graph, and eventually an evaluation context once those references are available.
+
+This theortically simple process is deeply complicated by the module dependency tree and expansion therein. The graph is dynamically modified due to `for_each` and `count` being evaluated as their required references are made available. The majority of the logic in this process exists within the `tofu` and `lang` package and are somewhat tightly coupled.
```
Comment (Yantrio) [inline docs/static-evaluation.md:27]:
```suggestion
This theoretically simple process is deeply complicated by the module dependency tree and expansion therein. The graph is dynamically modified due to `for_each` and `count` being evaluated as their required references are made available. The majority of the logic in this process exists within the `tofu` and `lang` package which are somewhat tightly coupled.
```

File: docs/static-evaluation.md [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+
+For example, a module's `for_each` statement may require data from a resource: `for_each = resource.aws_s3_bucket.foo.tags`. Before it could be evaluated, the module must wait for "OpenTofu Resource Reference aws_s3_bucket.foo" to be available. This would be represented as a dependency edge between the module node and the specific resource node. The evaluation context would then include `{"resource": {"aws_s3_bucket": {"foo": {"tags": <provided value>}}}}`.
+
+Note: A common misconception is that modules are "objects". Modules more closely resemble "namespaces" and can cross reference each other's vars/outputs as long as there is no reference loop.
```
Comment (Yantrio) [inline docs/static-evaluation.md:31]:
```suggestion
> [!NOTE]
> A common misconception is that modules are "objects". However, modules more closely resemble "namespaces" and can cross reference each other's vars/outputs as long as there is no reference loop.
```

File: docs/static-evaluation.md [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+
+
+## Initial implementation
+
+As you can see above, the lack of a building and managing evaluation contexts during the config loading stage prevents any expressions with references from being evaluation. Only primitive types and expressions are allowed during that stage.
```
Comment (Yantrio) [inline docs/static-evaluation.md:36]:
```suggestion
As you can see above, the lack of a building and managing evaluation contexts during the config loading stage prevents any expressions with references from being evaluated. Only primitive types and expressions are allowed during that stage.
```

File: docs/static-evaluation.md [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+
+Implementing this initial concept will allow many of the Solutions below to be fully implemented.  However, there are some limitations due to module expansion which are worth considering.
+
+## Additional Complexity due to Module Expansion
+
+The concepts of `for_each` and `count` were grafted on to the codebase in a way that has added significant complexity and limitations. When a module block contains a `for_each` or `count` all of the nodes (resources/variables/locals/etc...) will be created multiple times, one copy per "instance".
+
+One common example is using different providers for different module instances:
```
Comment (Yantrio) [inline docs/static-evaluation.md:52]:
```suggestion
One common example of a limitation would be to use different providers for different module instances:
```

File: docs/static-evaluation.md [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+  }
+```
+
+### Static Context Design
+At the heart of the project lies an evaluation context, similar to what currently exist in the tofu and lang package. It must serve a similar purpose, but has some differing requirements.
+
+Any static evaluator must be able to:
+* Evaluate a hcl expression or block into a single cty value
+  - Provide detailed insight into why a given expression or block can not be turned into a cty value
+* Be scoped to a given path
```
Comment (Yantrio) [inline docs/static-evaluation.md:200]: Can you expand on what this means? is it around child modules?

File: docs/static-evaluation.md [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+
+This theoretically simple process is deeply complicated by the module dependency tree and expansion therein. The graph is dynamically modified due to `for_each` and `count` being evaluated as their required references are made available. The majority of the logic in this process exists within the `tofu` and `lang` package which are somewhat tightly coupled.
+
+For example, a module's `for_each` statement may require data from a resource: `for_each = resource.aws_s3_bucket.foo.tags`. Before it could be evaluated, the module must wait for "OpenTofu Resource Reference aws_s3_bucket.foo" to be available. This would be represented as a dependency edge between the module node and the specific resource node. The evaluation context would then include `{"resource": {"aws_s3_bucket": {"foo": {"tags": <provided value>}}}}`.
+
+> [!NOTE]
+> A common misconception is that modules are "objects". However, modules more closely resemble "namespaces" and can cross reference each other's vars/outputs as long as there is no reference loop.
```
Comment (ronnyorot) [inline docs/static-evaluation.md:45]: Out of curiosity, how does tofu handle a module reference loop situation?

Comment (cam72cam) [reply docs/static-evaluation.md:45]: It currently prints out two different error messages, which are not super clear:
```
╷
│ Error: Reference to undeclared resource
│ 
│   on main.tf line 4, in module "a":
│    4:         in = b.out
│ 
│ A managed resource "b" "out" has not been declared in the root module.
╵
╷
│ Error: Reference to undeclared resource
│ 
│   on main.tf line 9, in module "b":
│    9:         in = a.out
│ 
│ A managed resource "a" "out" has not been declared in the root module.
```

File: docs/static-evaluation.md [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+
+## Initial implementation
+
+As you can see above, the lack of a building and managing evaluation contexts during the config loading stage prevents any expressions with references from being evaluated. Only primitive types and expressions are allowed during that stage.
+
+By introducing the ability to build and manage evaluation contexts during config loading, we would open up the ability for *certain* references to be evaluated during the config loading process.
+
+For example, many users expect to be able to use `local` values within `module -> source` to simplify upgrades and DRY up their configuration. This is not currently possible as the value of `module -> source` *must* be known during the config loading stage and can not be deferred until graph evaluation.
+
+By utilizing Traversals/References, we can track what values are statically known throughout the config loading process. This will follow a similar pattern to the graph reference evaluation (with limitations) and may or may not re-use much of it's code.
```
[the remark this answers is not in the document]
Comment (cam72cam) [reply docs/static-evaluation.md:56]: They are explained above, now with links.

File: docs/static-evaluation.md [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+
+### Backend Configuration
+
+Once the core is implemented, this is probably the easiest solution to implement.
+
+Notes from initial prototyping:
+* The configs.Backend saves the config body during the config load and does not evaluate it
+* backendInitFromConfig() in command/meta_backend.go is what evaluates the body
+  - This happens before the graph is constructed / evaluated and can be considered an extension of the config loading stage.
+* We can stash a copy of the StaticContext in the configs.Backend and use it in backendInitFromConfig() to provide do the decoding.
```
Comment (Evi1Pumpkin) [inline docs/static-evaluation.md:315]:
```suggestion
* We can stash a copy of the StaticContext in the configs.Backend and use it in backendInitFromConfig() do the decoding.
```

Comment (Evi1Pumpkin) [reply docs/static-evaluation.md:315]: Also, in "to do the decoding", you mean to evaluate the references instead, right?

File: docs/static-evaluation.md [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+
+Existing testing within OpenTofu is fragmented and more sparse than we would like. Additional test coverage will be needed before, during and after many of the following items.
+
+Code coverage should be inspected before refactoring of a component is undertaken to guide the additional test coverage required. We are not aiming for 100%, but should use it as a tool to understand our current testing.
+
+A comprehensive guide on e2e testing should be written, see #1536.
+
+## Unknowns:
+### Providers variables
+Providers may have configuration that depends on variables and dynamic values, such as resources from other providers. There is a odd workaround within the internal/tofu package where variable values may be requested during the graph building phase. This is an odd hack and may need to be reworked for the providers iteration above.
```
Comment (Evi1Pumpkin) [inline docs/static-evaluation.md:350]: Can you add a link to where it exists in the code? Also, a link to the documentation discussing it if it exists.

File: docs/static-evaluation.md [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+### Provider Iteration
+
+In [#300](https://github.com/opentofu/opentofu/issues/300), users describe how supporting for_each for configuring providers will allow much DRYer and simpler configurations.
+
+> [!Note]
+> Please read [Provider References](provider-references.md) before diving into this section!
+
+#### Proposed Changes
+
+The first change is to support static variables in the provider config block.  This can then be extended to support for_each/count and be expanded at the end of the `config.NewModule()` function, similar to how module.ProviderLocalNames is generated.  This piece is fairly straightforward and can be done relatively easily.
```
Comment (Evi1Pumpkin) [inline docs/static-evaluation.md:255]: Can you please provide an example for this case? I think I got it, but it'll reassure me if I'll see the TF code describing this scenario.

File: docs/static-module-expansion.md [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+        name = each.key
+        providers {
+          aws = provider.aws[each.value]
+        }
+}
+```
+
+Each "instance" of "mod" will need it's own provider configuration, which is currently only specified as part of "mod" and not any "mod instance". This problem becomes more complex and problematic when considering nested modules.
+
+To solve this, modules which have static for_each and count expressions must be expanded at the config layer. This must happen before the graph building, transformers, and walking.
```
Comment (RLRabinowitz) [inline docs/static-module-expansion.md:20]: Maybe I do not understand the meaning behind this, but I don't think we can **always** expand the module at config layer. We still have to support cases where the `for_each` is only evaluated in plan walk

File: docs/static-evaluation.md [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+
+## Unknowns:
+### Providers variables
+Providers may have configuration that depends on variables and dynamic values, such as resources from other providers. There is a odd workaround within the internal/tofu package where variable values may be requested during the graph building phase. This is an odd hack and may need to be reworked for the providers iteration above.
+### Core functions
+Do we want to support the core OpenTofu functions in the static evaluation context? Probably as it would be fairly trivial to hook in.
+### Provider functions
+Do we want to support provider functions during this static evaluation phase? I suspect not, without a good reason as the development costs may be significant with minimal benefit. It is trivial to detect someone attempting to use a provider function in an expression/body and to mark the expression result as dynamic.
+### Module Expansion Disk Copy
+As described in #issue, large projects may incur a large cost to build a directory for every remote module.  *If* we are expanding modules in a static context when possible, that implies that we will also be building a directory for every remote module instance.
```
Comment (ollevche) [inline docs/static-evaluation.md:383]: #issue?

File: docs/static-evaluation.md [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+  - Would likely require some ergonomic gymnastics depending on scale of refactoring
+
+This will need to be investigated and roughly prototyped, but all solutions should fit a similar enough interface to not block development of dependent tasks. We should design the interface first, based on the requirements of the initial prototype.
+
+
+## Solutions
+### Module Sources
+Module sources must be known at init time as they are downloaded and collated into .terraform/modules. This can be implemented by inspecting the source hcl.Expression using the static evaluator scoped to the current module.
+
+This is relatively straight forward once the core is implemented, but will require some more in-depth changes to support for_each/count later on.
```
[the remark this answers is not in the document]
Comment (RLRabinowitz) [reply docs/static-evaluation.md:240]: nvm I see that we do have such considerations in the `static-module-expansion.md`

File: docs/static-module-expansion.md (hunk 14 above)
Comment (cam72cam) [reply docs/static-module-expansion.md:20]: That's sort of what I meant by "static for_each and count expressions", but could reword it as "for_each and count expressions that can be statically evaluated"

File: docs/static-evaluation.md [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+  }
+```
+
+### Static Context Design
+At the heart of the project lies an evaluation context, similar to what currently exist in the tofu and lang package. It must serve a similar purpose, but has some differing requirements.
+
+Any static evaluator must be able to:
+* Evaluate a hcl expression or block into a single cty value
+  - Provide detailed insight into why a given expression or block can not be turned into a cty value
+* Be constructed with variables derived from a parent static context
```
[the remark this answers is not in the document]
Comment (ollevche) [reply docs/static-evaluation.md]: Resolved in [8f763bc](https://github.com/opentofu/opentofu/pull/1649/commits/8f763bc102050bada4e72676c69ea7ba6cd558ce)

File: docs/static-evaluation.md [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+
+## Plan of Attack
+
+Between the initial implementation, the simple solutions, and the provider complexity, we are talking about a significant amount of work likely spread across multiple releases.
+
+We can not take the approach of hacking on a feature branch for months or freezing all related code. It's unrealistic and unfair to other developers.
+
+Instead, we can break this work into smaller discrete and testable components, some of which may be easy to work on in parallel.
+
+If we design an interface for the static evaluator and wire a noop implementation through the config package, work on all of the major solutions can be started. Those solutions will then become functional as pieces of the static evaluator are implemented.
```
[the remark this answers is not in the document]
Comment (ollevche) [reply docs/static-evaluation.md]: Resolved in [8f763bc](https://github.com/opentofu/opentofu/pull/1649/commits/8f763bc102050bada4e72676c69ea7ba6cd558ce)

File: docs/static-evaluation.md [hunk 19]
```diff
@@ -0,0 +1,388 @@
...
+        name = each.key
+        providers {
+          aws = provider.aws[each.value]
+        }
+}
+```
+
+As the provider requirements are baked into the module itself, the multiple "instances" don't have any concept of providers per instance. This becomes even more complex when you consider that these providers might be passed through a complex tree of modules before they are directly used.
+
+There are two potential solutions detailed below in [Provider Iteration](#provider-iteration), each with their own tradeoffs.  Depending on the decision made there, we may not allow each.key/vaue and count to be used in module sources (at least for now).
```
[the remark this answers is not in the document]
Comment (ollevche) [reply docs/static-evaluation.md]: Resolved in [8f763bc](https://github.com/opentofu/opentofu/pull/1649/commits/8f763bc102050bada4e72676c69ea7ba6cd558ce)

File: docs/static-evaluation.md [hunk 20]
```diff
@@ -0,0 +1,446 @@
...
+
+As the provider requirements are baked into the module itself, the multiple "instances" don't have any concept of providers per instance. This becomes even more complex when you consider that these providers might be passed through a complex tree of modules before they are directly used.
+
+There are two potential solutions detailed below in [Provider Iteration](#provider-iteration), each with their own tradeoffs.
+
+A related problem is using `each.key/value` or `count` in module sources.  The sources don't have any concept of "module instances" from `for_each/count` and have similar limitations to provider iteration.  This a much lower priority item that is likely not going to be considered for the foreseeable future. TODO reference module expansion document.
+
+## Plan of Attack
+
+Between the initial implementation, the simple Solutions, and the provider complexity, we are talking about a significant amount of work likely spread across multiple releases.
```
Comment (siddharthasonker95) [inline docs/static-evaluation.md:128]: If this feature will have multiple releases, can a structured plan be created to detail which features will be rolled out in each release?

Comment (cam72cam) [reply docs/static-evaluation.md:128]: That's a good question.  I think I need to re-structure this RFC based on the conversations that have been happening around the new RFC process PR.  I think it is actually going to be two different RFCs in the end to properly scope the work.

File: docs/static-evaluation.md [hunk 21]
```diff
@@ -0,0 +1,446 @@
...
+# main.tf
+module "mod" {
+        for_each = {"us" = "first", "eu" = "second"}
+        source = "./mod"
+        name = each.key
+        providers {
+          aws = provider.aws[each.value] # Currently only hard-coded values are supported here
+        }
+}
+```
```
[the remark this answers is not in the document]
Comment (cam72cam) [reply docs/static-evaluation.md:118]: Yes

File: docs/static-evaluation.md [hunk 22]
```diff
@@ -0,0 +1,527 @@
...
+
+If we design an interface for the static evaluator and wire a noop implementation through the config package, work on all of the major solutions can be started. Those solutions will then become functional as pieces of the static evaluator are implemented.
+
+Additionally, the module iteration work can likely be broken into stages, implementing changes package by package and keeping around an address translation layer (detailed below).
+
+With this piece by piece approach, we can also add testing before, during, and after each component is added/modified.
+
+The OpenTofu core team should be the ones to do the majority of the core implementation and the module expansion work.  If community members are interested, many of the solutions are isolated and well defined enough for them to be worked on independently of the core team.
+
+## Progress Overview:
```
[the remark this answers is not in the document]
Comment (cam72cam) [reply docs/static-evaluation.md:140]: This will be removed as this is translated to the new RFC layout

File: rfc/20240513-static-evaluation.md [hunk 23]
```diff
@@ -0,0 +1,549 @@
...
+
+Performing an action in OpenTofu (init/plan/apply/etc...) takes the following steps (simplified):
+
+A [command](https://github.com/opentofu/opentofu/blob/290fbd66d3f95d3fa413534c4d5e14ef7d95ea2e/internal/command/init.go#L193) in the command package is created based on the CLI arguments and is executed.  It performs the following actions:
+
+#### Parse and Load the configuration modules
+
+Starting at the root module (current directory), a `config.Config` structure is created. This structure is the root node of a tree representing all of the module calls (`module {}`) that make up the project. Each node in the tree contains a `config.Module` and a `addrs.Module` path.
+
+The tree is built by: installing a module's source, loading the module, inspecting a the module calls, recursing in a depth first pattern.
```
Comment (ollevche) [inline rfc/20240513-static-evaluation.md:297]:
```suggestion
The tree is built by: installing a module's source, loading the module, inspecting the module calls, recursing in a depth first pattern.
```

File: rfc/20240513-static-evaluation.md [hunk 24]
```diff
@@ -0,0 +1,549 @@
...
+
+This will be described in the [Provider Evaluation RFC](20240513-static-evaluation-providers.md) due to expansion complexity.
+
+#### Blockers
+
+Existing testing within OpenTofu is fragmented and more sparse than we would like. Additional test coverage will be needed before, during and after each stage of implementation.
+
+Code coverage should be inspected before refactoring of a component is undertaken to guide the additional test coverage required. We are not aiming for 100%, but should use it as a tool to understand our current testing.
+
+A comprehensive guide on e2e testing should be written, see https://github.com/opentofu/opentofu/issues/#1536.
```
Comment (ollevche) [inline rfc/20240513-static-evaluation.md:505]:
```suggestion
A comprehensive guide on e2e testing should be written, see https://github.com/opentofu/opentofu/issues/1536.
```

File: rfc/20240513-static-evaluation.md [hunk 25]
```diff
@@ -0,0 +1,549 @@
...
+terraform {
+  backend "somebackend" {
+    region = local.region
+    key = var.key
+    key_check = local.key_check
+  }
+}
+```
+
+When first running `tofu init`, two errors will be produced:
```
Comment (Evi1Pumpkin) [inline rfc/20240513-static-evaluation.md:56]: If I run a plan/apply and don't configure a variable value, tofu asks me to provide the variable's value in the command line.
Maybe we'll also want the same behavior for the `init` command?

File: rfc/20240513-static-evaluation.md [hunk 26]
```diff
@@ -0,0 +1,549 @@
...
+variable "accesskey" {
+    type = string
+}
+
+terraform {
+  backend "somebackend" {
+    accesskey = var.accesskey
+  }
+}
+```
```
Comment (Yantrio) [inline rfc/20240513-static-evaluation.md:17]: this example here now feels a bit disjointed to the above and below statements.

File: rfc/20240513-static-evaluation.md (hunk 25 above)
Comment (Yantrio) [reply rfc/20240513-static-evaluation.md:56]: I think we should expand the explaination on line 60 below to cover this too.

Comment (RLRabinowitz) [reply rfc/20240513-static-evaluation.md:56]: I don't know if doing so would be easy, as we'd probably need specific code for backend configuration references + module source references, etc.
But it is something that is very worth considering (maybe not for the first phase)

Comment (cam72cam) [reply rfc/20240513-static-evaluation.md:44]: Updated in [26374cd](https://github.com/opentofu/opentofu/pull/1649/commits/26374cd5c7309128052eb075036589dfcbc71671)

File: docs/provider-references.md [hunk 27]
```diff
@@ -0,0 +1,352 @@
...
+
+## Language References
+
+### History and Addressing of Providers
+
+Provider references and configuration have an interesting history, which leads to the system we have today. Note: some of this history has been summarized or omitted for clarity.
+
+#### Provider Type
+
+Prior to v0.10.0, providers were built directly into the binary and not released/versioned separately.  They only had a single identifier, which we now call "Provider Type".
```
Comment (Yantrio) [inline docs/provider-references.md:33]:
```suggestion
Prior to Terraform v0.10.0, providers were built directly into the binary and not released/versioned separately.  They only had a single identifier, which we now call "Provider Type".
```

File: docs/provider-references.md [hunk 28]
```diff
@@ -0,0 +1,352 @@
...
+  }
+}
+```
+
+
+### Representation in State
+
+Resources in the state file note the `addrs.Provider` required to modify them. Due to the structure, all instances (for_each/count) of a resource must use the same provider.
+
+Note: This section should be expanded with examples.
```
Comment (Yantrio) [inline docs/provider-references.md:327]: we should also mention that the json representation of the statefile is different to the actual statefile.

File: rfc/20240513-static-evaluation-providers.md [hunk 29]
```diff
@@ -0,0 +1,356 @@
...
+
+#### Provider Alias Mappings
+
+The following understanding of the current OpenTofu code may be incorrect or incomplete. It is a wild mix of legacy patterns and fallbacks that is hard to reason about. It is based on [Provider References](../docs/provider-references.md#Provider-Workflow)
+
+Providers and their aliases are:
+* Fully known at init/config time
+* Hacked into the graph via ProviderTransformers
+* Attached to *Unexpanded* modules/resources in the graph
+* Linked to *Unexpanded* resources in the graph.
```
Comment (Yantrio) [inline rfc/20240513-static-evaluation-providers.md:215]: Let's not use charged language here.
```suggestion
The following understanding of the current OpenTofu code may be incorrect or incomplete. It is a mix of legacy patterns and fallbacks that is hard to reason about. It is based on [Provider References](../docs/provider-references.md#Provider-Workflow)

Providers and their aliases are:
* Fully known at init/config time
* Added into the graph via ProviderTransformers
* Attached to *Unexpanded* modules/resources in the graph
* Linked to *Unexpanded* resources in the graph.
```

File: rfc/20240513-static-evaluation-providers.md [hunk 30]
```diff
@@ -0,0 +1,356 @@
...
+
+Let's deconstruct each of these challenges individually:
+
+##### Providers through Init/Configuration:
+
+Each configs.Module contains fields which define how the module understands it's local provider references and provider configurations. As defined in Expansion above, we can use the StaticContext to fully build out these maps.
```
Comment (Yantrio) [inline rfc/20240513-static-evaluation-providers.md:221]:
```suggestion
Each `configs.Module` contains fields which define how the module understands it's local provider references and provider configurations. As defined in Expansion above, we can use the StaticContext to fully build out these maps.
```

**Review by Yantrio — CHANGES_REQUESTED** [review]
Some small nitpicks

File: docs/provider-references.md (hunk 27 above)
Comment (cam72cam) [reply docs/provider-references.md:33]: I tried to keep this fairly agnostic, what's the advantage of mentioning Terraform here?

File: docs/provider-references.md (hunk 28 above)
Comment (cam72cam) [reply docs/provider-references.md:327]: cd6ba01ea5

**Review by Yantrio — APPROVED** [review]

**Review by ollevche — APPROVED** [review]

File: rfc/20240513-static-evaluation-providers.md [hunk 31]
```diff
@@ -0,0 +1,356 @@
...
+
+Resources and Modules can be queried about which provider they reference. Due to legacy / implicit reasons, it is a bit of a complex question. The Resource and Module structures contain `configs.ProviderConfigRefs`, which is a config-friendly version of `addrs.LocalProviderConfig` (includes ranges).
+
+Resolving these provider references is a bit tricky and is a split responsibility between the ProviderTransformers in the graph and [validateProviderConfigs](https://github.com/opentofu/opentofu/blob/main/internal/configs/provider_validation.go#L292). This will require surgical changes to how provider configs are tracked between modules. Different approaches will need to be evaluated by the team and prototyped, likely with some refactoring beforehand.
+
+##### Providers in the graph
+
+As previously mentioned, the ProviderTransformers are tasked with inspecting the providers defined in configuration and attatching them to resources.
```
Comment (siddharthasonker95) [inline rfc/20240513-static-evaluation-providers.md:231]:
```suggestion
As previously mentioned, the ProviderTransformers are tasked with inspecting the providers defined in configuration and attaching them to resources.
```

**Review by siddharthasonker95 — APPROVED** [review]
tiny nitpick. overall LGTM!

File: rfc/20240513-static-evaluation-providers.md (hunk 31 above)
Comment (cam72cam) [reply rfc/20240513-static-evaluation-providers.md]: Fixed in [5019322](https://github.com/opentofu/opentofu/pull/1649/commits/50193220bff1e12c53f86b1c2f223478a9ccc81e)

**Comment by cam72cam** [comment]
I'm considering the prior approvals valid due to only changing spelling / grammar and updating a document that is not part of what will be implemented.

**Pull request merged**